workflow/executor/executor.go

```diff
@@ -802,7 +802,11 @@ func (we *WorkflowExecutor) FinalizeOutput(ctx context.Context) {
 			log.WithError(err).Warn("failed to patch task result, falling back to legacy/insecure pod patch, see https://argo-workflows.readthedocs.io/en/latest/workflow-rbac/")
 			// Only added as a backup in case LabelKeyReportOutputsCompleted could not be set
 			err = we.AddAnnotation(ctx, common.AnnotationKeyReportOutputsCompleted, "true")
+			if err != nil {
+				return err
+			}
 		}
+		err = we.RemoveFinalizer(ctx, common.FinalizerTaskResultStatus)
```

Comment:
The executor removes the new `/taskresult` finalizer here after the WorkflowTaskResult has been updated. If the executor is interrupted before this point is reached, will the finalizer ever be removed? As far as I understand, the controller does not currently do anything with it. 

If the controller were to remove the finalizer, it would also need to make sure that the associated WorkflowTaskResult is left in an appropriate state, as the executor is not guaranteed to have done so.

Reply:
Yes, does this PR address #12993? That refers specifically to non-graceful shutdown, which I'm assuming is the case in which a SIGKILL was received, preventing all of the usual graceful shutdown logic which would enable the `workflows.argoproj.io/report-outputs-completed` to be set to true.

Reply:
Yes，it address #12993，Avoid accidental deletion that causes taskresult to not be updated to completed

Reply:
I think the distinction is between "pod deletion" and "main container receiving a SIGKILL". A container can receive a SIGKILL preventing it from terminating gracefully even while the Pod is left up, right? 

I know there are different potential of this issue:
1. The Controller can issue a [SIGKILL](https://github.com/argoproj/argo-workflows/blob/main/workflow/executor/executor.go#L106) to the Container if it doesn't shut down normally with a SIGTERM.
2. Node deprovisioning
3. Node pressure eviction

I'm not sure how the latter 2 work. Do they send a SIGKILL to the container first? Do they simply delete the Pods, and does that simply cause a SIGKILL, a SIGTERM, or a SIGINT to be sent?